Mathlib/Data/Fin/Tuple/Basic.lean

```diff
@@ -284,21 +287,119 @@ theorem range_cons {α : Type _} {n : ℕ} (x : α) (b : Fin n → α) :
-  fun i ↦
-  if h : (i : ℕ) < m then u ⟨i, h⟩
-  else v ⟨(i : ℕ) - m, (tsub_lt_iff_left (le_of_not_lt h)).2 (ho ▸ i.isLt)⟩
+section Append
+
+/-- Append a tuple of length `m` to a tuple of length `n` to get a tuple of length `m + n`.
+This is a non-dependent version of `fin.add_cases`. -/
+def append {α : Type _} (a : Fin m → α) (b : Fin n → α) : Fin (m + n) → α :=
+  @Fin.addCases _ _ (fun _ => α) a b
 #align fin.append Fin.append
 
 @[simp]
-theorem fin_append_apply_zero {α : Type _} {o : ℕ} (ho : o + 1 = m + 1 + n) (u : Fin (m + 1) → α)
-    (v : Fin n → α) : Fin.append ho u v 0 = u 0 := by
-  rw [append]
-  simp
-#align fin.fin_append_apply_zero Fin.fin_append_apply_zero
+theorem append_left {α : Type _} (u : Fin m → α) (v : Fin n → α) (i : Fin m) :
+    append u v (Fin.castAdd n i) = u i :=
+  addCases_left _ _ _
+#align fin.append_left Fin.append_left
+
+@[simp]
+theorem append_right {α : Type _} (u : Fin m → α) (v : Fin n → α) (i : Fin n) :
+    append u v (natAdd m i) = v i :=
+  addCases_right _ _ _
+#align fin.append_right Fin.append_right
+
+theorem append_right_nil {α : Type _} (u : Fin m → α) (v : Fin n → α) (hv : n = 0) :
+    append u v = u ∘ Fin.cast (by rw [hv, add_zero]) := by
+  refine' funext (Fin.addCases (fun l => _) fun r => _)
+  · rw [append_left, Function.comp_apply]
+    refine' congr_arg u (Fin.ext _)
+    simp
+  · exact (Fin.cast hv r).elim0'
+#align fin.append_right_nil Fin.append_right_nil
+
+@[simp]
+theorem append_elim0' {α : Type _} (u : Fin m → α) :
+    append u Fin.elim0' = u ∘ Fin.cast (add_zero _) :=
+  append_right_nil _ _ rfl
+#align fin.append_elim0' Fin.append_elim0'
+
+theorem append_left_nil {α : Type _} (u : Fin m → α) (v : Fin n → α) (hu : m = 0) :
+    append u v = v ∘ Fin.cast (by rw [hu, zero_add]) := by
+  refine' funext (Fin.addCases (fun l => _) fun r => _)
+  · exact (Fin.cast hu l).elim0'
+  · rw [append_right, Function.comp_apply]
+    refine' congr_arg v (Fin.ext _)
+    simp [hu]
+#align fin.append_left_nil Fin.append_left_nil
+
+@[simp]
+theorem elim0'_append {α : Type _} (v : Fin n → α) :
+    append Fin.elim0' v = v ∘ Fin.cast (zero_add _) :=
+  append_left_nil _ _ rfl
+#align fin.elim0'_append Fin.elim0'_append
+
+theorem append_assoc {p : ℕ} {α : Type _} (a : Fin m → α) (b : Fin n → α) (c : Fin p → α) :
+    append (append a b) c = append a (append b c) ∘ Fin.cast (add_assoc _ _ _) := by
+  ext i
+  rw [Function.comp_apply]
+  refine' Fin.addCases (fun l => _) (fun r => _) i
+  · rw [append_left]
+    refine' Fin.addCases (fun ll => _) (fun lr => _) l
+    · rw [append_left]
+      simp [castAdd_castAdd]
+    · rw [append_right]
+      simp [castAdd_natAdd]
+  · rw [append_right]
+    simp [← natAdd_natAdd]
+#align fin.append_assoc Fin.append_assoc
+
+end Append
+
+section Repeat
+
+/-- Repeat `a` `m` times. For example `Fin.repeat 2 ![0, 3, 7] = ![0, 3, 7, 0, 3, 7]`. -/
+-- Porting note: removed @[simp]
```

Comment:
What went wrong if you kept it?

Reply:
"simp simplifies LHS" at https://github.com/leanprover-community/mathlib4/actions/runs/3947414172/jobs/6756204194

Reply:
In lean 3, `@[simp]` on an equation compiler definition tagged each of the equations with `simp`.

If that's not the case any more, then we should add a new `repeat_apply` lemma with the same statement.

Reply:
[Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/.40.5Bsimp.5D.20on.20equation.20compiler.20definitions/near/322022846)

Reply:
For now, let's add `repeat_apply` and include `leanprover/lean4#2042` in the porting note.